### PR TITLE
feat(ci): Add coverage run for invoke task and set them aside go tests

### PR DESCRIPTION
**Pull request by chouetz**
### What does this PR do?
Run invoke unit tests in dedicated jobs. Will add coverage afterwards

### Motivation
Invoke tasks are a different workload in the datadog-agent repo.
Their UT can run independently and having a dedicated workflow (and coverage) would give rapid feedback for devx team when touching those

### Additional Notes
I will probably create a side workflow for kitchen.invoke-unit-tests
We could easily use [coverage-scope](https://github.com/marketplace/actions/coverage-scope) action to enforce a gating on coverage for invoke tasks

### CODE REVIEW

File: .gitlab/source_test/tooling_unit_tests.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  rules:
+    - !reference [.on_invoke_tasks_changes]
+  script:
+    - source /root/.bashrc
+    - python3 -m pip install -r tasks/libs/requirements-github.txt
+    - inv -e invoke-unit-tests
+
+kitchen_invoke_unit_tests:
+  stage: source_test
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64$DATADOG_AGENT_BUILDIMAGES_SUFFIX:$DATADOG_AGENT_BUILDIMAGES
```
Comment (iglendd) [inline .gitlab/source_test/tooling_unit_tests.yml:17]: Would deb image work for windows kitchen tests?

Comment (chouetz) [reply .gitlab/source_test/tooling_unit_tests.yml:17]: In this case yes as it's testing the invoke tasks of kitchen (so the python code only). I assume there is few risk of misalignment of this between the python version we would have on the images (that should be the same by the way)

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -1352,3 +1352,17 @@ workflow:
...
+      paths:
+        - tasks/**/*
+      compare_to: main # TODO: use a variable, when this is supported https://gitlab.com/gitlab-org/gitlab/-/issues/369916
+
+.on_kitchen_invoke_tasks_changes:
+  - !reference [.except_mergequeue]
+  - changes:
+      paths:
+        - test/kitchen/tasks/**/*
+      compare_to: main # TODO: use a variable, when this is supported https://gitlab.com/gitlab-org/gitlab/-/issues/369916
```
Comment (CelianR) [inline .gitlab-ci.yml:1368]: Can you add a trailing line feed ?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -1352,3 +1352,17 @@ workflow:
 .on_fakeintake_changes_on_main:
   - <<: *on_fakeintake_changes
     if: $CI_COMMIT_BRANCH == "main"
+
+.on_invoke_tasks_changes:
+  - !reference [.except_mergequeue]
+  - changes:
+      paths:
+        - tasks/**/*
+      compare_to: main # TODO: use a variable, when this is supported https://gitlab.com/gitlab-org/gitlab/-/issues/369916
```
Comment (CelianR) [inline .gitlab-ci.yml:1361]:
```suggestion
    paths:
      - tasks/**/*
    compare_to: main # TODO: use a variable, when this is supported https://gitlab.com/gitlab-org/gitlab/-/issues/369916
```

File: .gitlab-ci.yml (hunk 2 above)
Comment (CelianR) [inline .gitlab-ci.yml:1368]:
```suggestion
    paths:
      - test/kitchen/tasks/**/*
    compare_to: main # TODO: use a variable, when this is supported https://gitlab.com/gitlab-org/gitlab/-/issues/369916
```

File: .gitlab/source_test/tooling_unit_tests.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  tags: ["arch:amd64"]
+  needs: []
+  rules:
+    - !reference [.on_kitchen_invoke_tasks_changes]
+  script:
+    - source /root/.bashrc
+    - python3 -m pip install -r tasks/libs/requirements-github.txt
+    - pushd test/kitchen
+    - inv -e kitchen.invoke-unit-tests
+    - popd
```
Comment (CelianR) [inline .gitlab/source_test/tooling_unit_tests.yml:27]: Can you add a trailing line feed ?

**Review by CelianR — APPROVED** [review]
Some style changes

File: .gitlab/source_test/tooling_unit_tests.yml (hunk 4 above)
Comment (chouetz) [reply .gitlab/source_test/tooling_unit_tests.yml:27]: done

File: .gitlab-ci.yml (hunk 2 above)
Comment (chouetz) [reply .gitlab-ci.yml:1368]: done

File: .gitlab-ci.yml (hunk 3 above)
Comment (chouetz) [reply .gitlab-ci.yml:1361]: This is the current indentation use elsewhere in the file, I'll stick to it right now

File: .gitlab-ci.yml (hunk 2 above)
Comment (chouetz) [reply .gitlab-ci.yml:1368]: Same as previous

File: .gitlab/source_test/linux.yml [hunk 5]
```diff
@@ -28,13 +28,6 @@
   script:
     - !reference [.retrieve_linux_go_tools_deps]
     - inv -e install-tools
-    - python3 -m pip install -r tasks/libs/requirements-github.txt
-    - inv -e invoke-unit-tests
-    - pushd test/kitchen
-    - inv -e kitchen.invoke-unit-tests
-    - popd
-    - FAST_TESTS_FLAG=""
-    - if [[ "$FAST_TESTS" == "true" ]]; then FAST_TESTS_FLAG="--only-impacted-packages"; fi
```
Comment (KSerrania) [inline .gitlab/source_test/linux.yml:37]: Is this expected? It sounds like this PR disables the fast tests features introduced in https://github.com/DataDog/datadog-agent/pull/23239.

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -1424,3 +1424,17 @@ workflow:
       FAST_TESTS: "false"
   - variables:
       FAST_TESTS: "true"
+
+.on_invoke_tasks_changes:
+  - !reference [.except_mergequeue]
+  - changes:
+      paths:
+        - tasks/**/*
+      compare_to: main # TODO: use a variable, when this is supported https://gitlab.com/gitlab-org/gitlab/-/issues/369916
```
Comment (KSerrania) [inline .gitlab-ci.yml:1433]: Are these supposed to run on `main`? I'd say yes, like all our other unit tests, but that is not what the rules are enforcing right now.

**Review by KSerrania — CHANGES_REQUESTED** [review]
Requesting changes to block the merge queue.

File: .gitlab/source_test/linux.yml (hunk 5 above)
Comment (chouetz) [reply .gitlab/source_test/linux.yml:37]: Wow no this is not expected I completely failed my rebase,  many thanks

File: .gitlab-ci.yml (hunk 6 above)
Comment (chouetz) [reply .gitlab-ci.yml:1433]: Do you mean `running every time on main` versus `running only when tasks are changed`?
If yes I'm not sure the first one is mandatory (at least that's what I tried to write), if not I probably misunderstood your comment

**Review by KSerrania — APPROVED** [review]

**Review by clarkb7 — APPROVED** [review]

**Pull request merged**